read-fonts/src/tables/postscript/charstring.rs

```diff
@@ -1,203 +1,644 @@
+    fn close(&mut self);
 }
 
-impl Operator {
-    /// Creates an operator from the given opcode.
-    fn from_opcode(opcode: u8) -> Option<Self> {
-        use Operator::*;
-        Some(match opcode {
-            1 => HStem,
-            3 => VStem,
-            4 => VMoveTo,
-            5 => RLineTo,
-            6 => HLineTo,
-            7 => VLineTo,
-            8 => RrCurveTo,
-            10 => CallSubr,
-            11 => Return,
-            14 => EndChar,
-            15 => VariationStoreIndex,
-            16 => Blend,
-            18 => HStemHm,
-            19 => HintMask,
-            20 => CntrMask,
-            21 => RMoveTo,
-            22 => HMoveTo,
-            23 => VStemHm,
-            24 => RCurveLine,
-            25 => RLineCurve,
-            26 => VvCurveTo,
-            27 => HhCurveTo,
-            29 => CallGsubr,
-            30 => VhCurveTo,
-            31 => HvCurveTo,
-            _ => return None,
-        })
+/// Command sink that sends the results of charstring evaluation to a [Pen].
+pub struct PenSink<'a, P>(&'a mut P);
+
+impl<'a, P> CommandSink for PenSink<'a, P>
+where
+    P: Pen,
+{
+    fn move_to(&mut self, x: Fixed, y: Fixed) {
+        self.0.move_to(x.to_f64() as f32, y.to_f64() as f32);
+    }
+
+    fn line_to(&mut self, x: Fixed, y: Fixed) {
+        self.0.line_to(x.to_f64() as f32, y.to_f64() as f32);
+    }
+
+    fn curve_to(&mut self, cx0: Fixed, cy0: Fixed, cx1: Fixed, cy1: Fixed, x: Fixed, y: Fixed) {
+        self.0.curve_to(
+            cx0.to_f64() as f32,
+            cy0.to_f64() as f32,
+            cx1.to_f64() as f32,
+            cy1.to_f64() as f32,
+            x.to_f64() as f32,
+            y.to_f64() as f32,
+        );
     }
 
-    /// Creates an operator from the given extended opcode.
-    ///
-    /// These are preceded by a byte containing the escape value of 12.
-    pub fn from_extended_opcode(opcode: u8) -> Option<Self> {
-        use Operator::*;
-        Some(match opcode {
-            34 => HFlex,
-            35 => Flex,
-            36 => HFlex1,
-            37 => Flex1,
-            _ => return None,
-        })
+    fn close(&mut self) {
+        self.0.close();
     }
 }
 
-/// Either a PostScript charstring operator or a (numeric) operand.
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
-pub enum Token {
-    Operator(Operator),
-    Operand(Number),
+/// Evaluates the given charstring and emits the resulting commands to the
+/// specified sink.
+///
+/// If evaluating a CFF2 charstring and the top-level table contains an
+/// item variation store, then `blend_state` must be provided.
+pub fn evaluate(
+    charstring_data: &[u8],
+    global_subrs: Index,
+    subrs: Option<Index>,
+    blend_state: Option<BlendState>,
+    sink: &mut impl CommandSink,
+) -> Result<(), Error> {
+    let mut evaluator = Evaluator::new(global_subrs, subrs, blend_state);
+    evaluator.evaluate(charstring_data, Fixed::ZERO, Fixed::ZERO, sink, 0)?;
```

Comment:
+1, this sticks out ot me too. A comment on evaluator might help.

Reply:
This is transient state. Moved most of these parameters to fields, but left `charstring_data` and `nesting_depth` as those are directly dependent on the recursive calls. The cursor doesn't fit as a field because it's only valid for one stack frame.

Split out operator processing to a separate method from general flow and number parsing.